### PR TITLE
Dataset preparation

### PR DESCRIPTION
**Pull request by btobab**
One-line command for image and label division, All you need to do is create a new folder dataset, place the image in the preimages directory, place the label in the Annotations directory, create two new folders images, labels, and finally modify the classes in the voc2txt file to their own category

### Summary
📂🐍📷

This pull request adds several scripts to the `prepare` folder that automate the process of preparing the dataset for YOLOv5 training. The scripts split the images and labels into train, val and test sets, convert the VOC format to YOLO format, and clean up the unnecessary files.

> _`prepare.sh` script_
> _splits and converts the data_
> _autumn leaves fall fast_

### Walkthrough
*  Add scripts to prepare the dataset for YOLOv5 training ([link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-31ba34c19b039fd45d658f132f8e6e96b95fadd6472a58cf1664ec802459a848R1-R10), [link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-3d62b198fe9741e052ec207dfe5e944c04f80ee23aee18b159b31c6b7a5f9dd5R1-R29), [link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-b56071e6c2146317525c83a6572cc547a848cdec3923d0dc5cd1b5843c0bc7b8R1-R20), [link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-542d636c1005cdc54e365fa271f13ef170b575b86e8e9037715b91a559dab42fR1-R23), [link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-5f7df3ce16c9d8a37113aee16597607281259b21607a25c477b1d243ca6dbf72R1-R48))
  * Split the image names into train, val and test lists and write them to text files using `split_images2txt.py` ([link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-3d62b198fe9741e052ec207dfe5e944c04f80ee23aee18b159b31c6b7a5f9dd5R1-R29))
  * Copy the images from the `preimages` folder to the `images` folder according to the split lists using `split_images_from_txt.py` ([link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-b56071e6c2146317525c83a6572cc547a848cdec3923d0dc5cd1b5843c0bc7b8R1-R20))
  * Convert the VOC format annotations to the YOLO format labels using `voc2txt.py` and write them to the `prelabels` folder ([link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-5f7df3ce16c9d8a37113aee16597607281259b21607a25c477b1d243ca6dbf72R1-R48))
  * Copy the labels from the `prelabels` folder to the `labels` folder according to the split lists using `split_labels_fromOne2N.py` ([link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-542d636c1005cdc54e365fa271f13ef170b575b86e8e9037715b91a559dab42fR1-R23))
  * Add commands to the `prepare.sh` script to run the above scripts and delete unnecessary files ([link](https://github.com/ultralytics/yolov5/pull/11151/files?diff=unified&w=0#diff-31ba34c19b039fd45d658f132f8e6e96b95fadd6472a58cf1664ec802459a848R1-R10))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
New data preparation scripts added to streamline dataset setup for YOLOv5 training.

### 📊 Key Changes
- 🆕 Added `prepare.sh` script to execute data preparation processes.
- 🖼️ Included `split_images2txt.py` for dividing images into train, validation, and test sets.
- ✂️ Created `split_images_from_txt.py` to organize images into corresponding directories.
- ↔️ Implemented `voc2txt.py` to convert VOC annotation format to YOLO text format.
- 🏷️ Developed `split_labels_fromOne2N.py` to distribute labels into train, val, and test directories.

### 🎯 Purpose & Impact
- 🛠️ These scripts prepare image datasets and annotations for training, validating, and testing in the YOLOv5 model, handling conversion and organization tasks.
- 👨‍💻 Users will benefit from a streamlined setup process, potentially increasing the efficiency and ease of initiating new model training sessions.
- 🚀 The impact will likely be felt by users requiring fast and reliable setup for various datasets, further enhancing the user experience with Ultralytics YOLOv5.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@btobab thanks for the PR! 

We have an `autosplit()` function for this purpose here:
https://github.com/ultralytics/yolov5/blob/5543b89466d072a9f8f2e31f8257a1ccc7f588e9/utils/dataloaders.py#L964-L988

**Comment by glenn-jocher** [comment]
copilot:all

**Comment by lightfate** [comment]
ChatGPT コードレビュー: ``prepare/prepare.sh``

 - 誤字があるので修正しましょう（connvert → convert）
- ファイルを削除する前に、ファイルが存在することを事前に確認する必要があります。ファイルが存在しない場合には、エラーが発生するため、処理が停止します。従って、ファイルが存在しなくてもエラーが発生しないように、「-f」オプションを付けて解決策を示しましょう。

**Comment by lightfate** [comment]
ChatGPT コードレビュー: ``prepare/split_images2txt.py``

 - コード全体の説明コメントを加えるとよい
- グローバルスコープの変数の命名がわかりにくいので、関数内で引数として渡すべき
- ファイルパスを直接記述せず、引数として渡す方が柔軟性がある
- パスの区切りをOSに依存しないよう、`os.path.join()`を使うとよい

**Comment by lightfate** [comment]
ChatGPT コードレビュー: ``prepare/split_images_from_txt.py``

 - `modes`変数に格納されているモード（`train`, `val`, `test`）の文字列操作は，（将来的にモード名が`train`, `valid`, `test`などに変更する可能性があるため）無名関数や定数として定義し，参照するべき
- `split_images_from_txt.py`が新規作成されたため，文書内の変数・モジュールインポート構成が不明瞭。循環輸送により，例えば`os.system()`メソッドは個別のユーティリティとして別モジュールに切り離した方がよい。
- 例外処理は具体的な例外を指定することを推奨する。今回は書き込み不可能なディレクトリにおける例外が発生するため，`FileNotFoundError`か `PermissionError`例外をキャッチすることを推奨します。

```python
import os
from typing import List
from tqdm import tqdm


class ImageDatasetSplitter:
    MODES = ["train", "val", "test"]
    IMAGE_DIR_PATH = "../dataset/images/"
    PREIMAGE_DIR_PATH = "../dataset/preimages/"

    def __init__(self):
        for mode in ImageDatasetSplitter.MODES:
            mode_path= os.path.join(ImageDatasetSplitter.IMAGE_DIR_PATH, mode)
            if not os.path.exists(mode_path):
                os.mkdir(mode_path)
                print(f"create {mode_path}")

    def split_dataset_to_images(self, mode: str, data: List[str]):
        for image_name in tqdm(data):
            preimage_file_path = os.path.join(
                ImageDatasetSplitter.PREIMAGE_DIR_PATH, f"{image_name}.jpg")
            if os.path.exists(preimage_file_path):
                try:
                    target_path = os.path.join(
                        ImageDatasetSplitter.IMAGE_DIR_PATH, mode, f"{image_name}.jpg")
                    os.system(f"cp {preimage_file_path} {target_path}")
                except (FileNotFoundError, PermissionError) as e:
                    print(f"writing {image_name} failed due to {e}")
            else:
                print(f"{preimage_file_path} is not exist")

```

**Comment by lightfate** [comment]
ChatGPT コードレビュー: ``prepare/split_labels_fromOne2N.py``

 - 1行目にファイルの説明がないので、docstring を書きましょう。
- `os.mkdir(os.path.join(mode_labels_path))`の `os.path.join`は不要です。また、これにより、不要な空のディレクトリが作成されることがあるため、何かしら中身を入れるべきです。
- `pre_label_path` が存在しない場合に受け付けるエラーがOSErrorになっていますが、FileNotFoundErrorにするべきです。
- printの内容を省略することができるので、詳細を記述するようにしましょう。ただし、Loggerを使用して、より詳細なログを出力するようにすると、より良いプラクティスと言えます。

**Comment by lightfate** [comment]
ChatGPT コードレビュー: ``prepare/voc2txt.py``

 - 関数や変数にわかりやすい名前をつけるように意識しましょう
- `('/dataset/Annotations/%s.xml' % (image_id)`のように文字列の結合に%演算子を使用する代わりにf-stringを使用すると、より可読性が高くなります
- 2つの空白行があるので削除しましょう。コード全体の見た目が良くなります

**Comment by glenn-jocher** [comment]
@lightfate thanks for the feedback on `prepare/voc2txt.py`!

You're absolutely right about the readability improvements. Here's an updated version of the code that incorporates your feedback:

```python
import os.path as osp


def create_directory_if_not_exists(directory: str):
    """Create directory if it doesn't exist."""
    if not osp.exists(directory):
        os.makedirs(directory)


def convert_voc_to_yolo(image_id: str, xml_file: str, class_dict: dict):
    with open(xml_file) as in_file:
        with open(f'../dataset/prelabels/{image_id}.txt', 'w') as out_file:
            for line in in_file:
                if '<name>' in line:
                    class_name = line.strip().split('>')[1].split('<')[0]
                    class_id = class_dict[class_name]
                if '<bndbox>' in line:
                    xmin = int(next(in_file).strip().split('>')[1].split('<')[0])
                    ymin = int(next(in_file).strip().split('>')[1].split('<')[0])
                    xmax = int(next(in_file).strip().split('>')[1].split('<')[0])
                    ymax = int(next(in_file).strip().split('>')[1].split('<')[0])

                    width, height = get_image_width_height(image_id, "jpg")
                    dw = 1./width
                    dh = 1./height
                    x = (xmin + xmax)/2.0
                    y = (ymin + ymax)/2.0
                    w = (xmax - xmin)*dw
                    h = (ymax - ymin)*dh

                    out_file.write(f"{class_id} {x} {y} {w} {h}\n")
```

This should be more readable and easy to understand. Let me know if you have any more feedback on this code or anything else!
